plugins/rocksdb/src/test/java/org/hyperledger/besu/plugin/services/storage/rocksdb/segmented/LayeredKeyValueStorageTest.java

```diff
@@ -0,0 +1,339 @@
+
+    List<Pair<byte[], byte[]>> resultList = result.collect(Collectors.toList());
+    assertEquals(1, resultList.size());
+    assertArrayEquals(key1, resultList.get(0).getKey());
+    assertArrayEquals(value1, resultList.get(0).getValue());
+  }
+
+  @Test
+  void shouldReturnLayerDataWhenParentIsEmpty() {
+    byte[] key1 = {1};
+    byte[] value1 = {10};
+
+    when(parentStorage.stream(segmentId)).thenReturn(Stream.empty());
+
+    var hashValueStore = createSegmentMap();
+    hashValueStore.get(segmentId).put(Bytes.wrap(key1), Optional.of(value1));
+    layeredKeyValueStorage = new LayeredKeyValueStorage(hashValueStore, parentStorage);
+
+    Stream<Pair<byte[], byte[]>> result = layeredKeyValueStorage.stream(segmentId);
+    List<Pair<byte[], byte[]>> resultList = result.toList();
+    assertEquals(1, resultList.size());
+    assertArrayEquals(key1, resultList.get(0).getKey());
+    assertArrayEquals(value1, resultList.get(0).getValue());
+  }
+
+  @Test
+  void shouldMergeParentAndLayerData() {
+    byte[] key1 = {1};
+    byte[] value1 = {10};
+    byte[] key2 = {2};
+    byte[] value2 = {20};
+    byte[] key3 = {3};
+    byte[] value3 = {30};
+
+    when(parentStorage.stream(segmentId))
+        .thenReturn(Stream.of(Pair.of(key1, value1), Pair.of(key3, value3)));
+
+    var hashValueStore = createSegmentMap();
+    hashValueStore.get(segmentId).put(Bytes.wrap(key2), Optional.of(value2));
+    layeredKeyValueStorage = new LayeredKeyValueStorage(hashValueStore, parentStorage);
+
+    Stream<Pair<byte[], byte[]>> result = layeredKeyValueStorage.stream(segmentId);
+
+    List<Pair<byte[], byte[]>> resultList = result.toList();
+    assertEquals(3, resultList.size());
+    assertArrayEquals(key1, resultList.get(0).getKey());
+    assertArrayEquals(value1, resultList.get(0).getValue());
+    assertArrayEquals(key2, resultList.get(1).getKey());
+    assertArrayEquals(value2, resultList.get(1).getValue());
+    assertArrayEquals(key3, resultList.get(2).getKey());
+    assertArrayEquals(value3, resultList.get(2).getValue());
+  }
+
+  @Test
+  void shouldPreferLayerDataOverParentDataForSameKey() {
+    byte[] key = {1};
+    byte[] parentValue = {10};
+    byte[] layerValue = {20};
+
+    when(parentStorage.stream(segmentId)).thenReturn(Stream.of(Pair.of(key, parentValue)));
+
+    var hashValueStore = createSegmentMap();
+    hashValueStore.get(segmentId).put(Bytes.wrap(key), Optional.of(layerValue));
+    layeredKeyValueStorage = new LayeredKeyValueStorage(hashValueStore, parentStorage);
+
+    Stream<Pair<byte[], byte[]>> result = layeredKeyValueStorage.stream(segmentId);
+
+    List<Pair<byte[], byte[]>> resultList = result.toList();
+    assertEquals(1, resultList.size());
+    assertArrayEquals(key, resultList.get(0).getKey());
+    // Layer value should be returned
+    assertArrayEquals(layerValue, resultList.get(0).getValue());
+  }
+
+  @Test
+  void shouldNotStreamKeyIfLayerKeyIsEmpty() {
```

Comment:
could you also create a test where `parent` also contains empty values to make sure they are also filtered out?

Reply:
https://github.com/hyperledger/besu/pull/7535#issuecomment-2317016065

I think it was in the test list that I proposed to add? Or is it another case?

Reply:
yeah, `shouldStreamWithMultipleLayersAndCurrentLayerOverridesWithValues` tests this scenario